src/crystal/event_loop/io_uring.cr

```diff
@@ -0,0 +1,483 @@
+      sqe.value.fd = fd
+      sqe.value.len = LibC::SHUT_RD # FIXME: add SHUT_WR too (?)
+    end
+  end
+
+  # internals
+
+  private def check_open(io)
+    raise IO::Error.new("Closed stream") if io.closed?
+  end
+
+  private def async_rw(opcode, io, slice, timeout, &)
+    loop do
+      res = async(opcode, timeout) do |sqe|
+        sqe.value.fd = io.fd
+        sqe.value.u1.off = -1
+        sqe.value.addr = slice.to_unsafe.address.to_u64!
+        sqe.value.len = slice.size
+      end
+      return res if res >= 0
+
+      check_open(io)
+      yield Errno.new(-res) unless res == -LibC::EINTR
+    end
+  end
+
+  private def async_poll(io, poll_events, timeout, &)
+    res = async(LibC::IORING_OP_POLL_ADD, timeout) do |sqe|
+      sqe.value.fd = io.fd
+      sqe.value.sflags.poll_events = poll_events | LibC::POLLERR | LibC::POLLHUP
+    end
+    check_open(io)
+    raise IO::TimeoutError.new(yield) if res == -LibC::ECANCELED
+  end
+
+  private def async_close(fd, &)
+    res = async_impl do |event|
+      @ring.reserve(2)
+
+      # linux won't interrupt pending operations on a file descriptor when it
+      # closes it, we thus first create an operation to cancel any pending
+      # operations; we don't attach that cancel operation to an event: handling
+      # the CQE for close is enough
+      cancel_sqe = @ring.unsafe_next_sqe
+      cancel_sqe.value.flags = LibC::IOSQE_IO_LINK | LibC::IOSQE_IO_HARDLINK | LibC::IOSQE_CQE_SKIP_SUCCESS
+      yield cancel_sqe
+      trace(cancel_sqe)
+
+      # then we setup the close operation
+      close_sqe = @ring.unsafe_next_sqe
+      close_sqe.value.opcode = LibC::IORING_OP_CLOSE
+      close_sqe.value.user_data = event.address.to_u64!
+      close_sqe.value.fd = fd
+      trace(close_sqe)
+    end
+
+    case res
+    when 0
+      # success
+    when -LibC::EINTR, -LibC::EINPROGRESS
+      # ignore
+    else
+      raise IO::Error.from_os_error("Error closing file", Errno.new(-res))
+    end
+  end
+
+  private def async_timeout(type : Event::Type, duration)
+    async_impl(type) do |event|
+      @ring.reserve(1)
+
+      sqe, ts = @ring.unsafe_next_sqe_with_timespec
+
+      ts.value.tv_sec = typeof(ts.value.tv_sec).new(duration.@seconds)
+      ts.value.tv_nsec = typeof(ts.value.tv_nsec).new(duration.@nanoseconds)
+
+      sqe.value.opcode = LibC::IORING_OP_TIMEOUT
+      sqe.value.user_data = event.address.to_u64!
+      sqe.value.addr = ts.address.to_u64!
+      sqe.value.len = 1
+      trace(sqe)
+    end
+  end
+
+  private def async(opcode, timeout = nil, &)
+    async_impl do |event|
+      @ring.reserve(timeout ? 2 : 1)
+
+      # configure the operation
+      op_sqe = @ring.unsafe_next_sqe
+      op_sqe.value.opcode = opcode
+      op_sqe.value.user_data = event.address.to_u64!
+      yield op_sqe
+
+      if timeout
+        # chain the above operation with the next one
+        op_sqe.value.flags = op_sqe.value.flags | LibC::IOSQE_IO_LINK
```

Comment:
In anticipation to @Sija suggesting to use `|=` ( :grin: ) ... NOPE! That would _copy_ the pointer value and then update the flag of the copy, leaving `op_sqe.value.flags` to its old value. It took me a while to identify & fix :hurtrealbad:

Reply:
I suppose that's a downside of expanding syntax sugar like `|=` and not having pointer dereference as a language feature (`pointer.value` is just a method call, implemented as a compiler primitive).